raco/rules.py

```diff
@@ -1041,3 +1043,55 @@ def check_partition_equality(op, representation):
     """
 
     return op.partitioning().hash_partitioned == frozenset(representation)
+
+
+class MoveDeBroadcast(Rule):
+
+    @staticmethod
+    def _copy_op(other):
+        try:
+            copy = other.__class__()
+        except Exception as e:
+            raise Exception("{}".format(other))
+
+        copy.copy(other)
+        return copy
+
+    def fire(self, expr):
+        if not isinstance(expr, algebra.DeBroadcast):
+            return expr
+
+        if expr.input.partitioning().broadcasted:
+            if isinstance(expr.input, algebra.ZeroaryOperator):
+                # tell the Scan/ScanTemp to scan only one replica
+                copy = MoveDeBroadcast._copy_op(expr.input)
+                copy._debroadcast = True
+                return copy
+            elif isinstance(expr.input, algebra.UnaryOperator):
+                copy = MoveDeBroadcast._copy_op(expr.input)
+                exprcopy = MoveDeBroadcast._copy_op(expr)
+                exprcopy.input = copy.input
+                copy.input = exprcopy
+                return copy
+            elif isinstance(expr.input, algebra.BinaryOperator):
```

Comment:
add comment that this case pushes a DeBroadcast op down one level on each branch